File: FirmwareGO/webapp/middleware/prometheus/CustomMetrics.go
package prometheus

import (
	"github.com/gin-gonic/gin"
	"github.com/prometheus/client_golang/prometheus"
	"strconv"
	"time"
)

// Custom metrics setup
var (
	httpRequestDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
		Name: "http_request_duration_seconds",
		Help: "Duration of HTTP requests in seconds.",
	}, []string{"method", "status_code", "path"})

	httpRequestCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "http_request_total",
		Help: "Total count of HTTP requests",
	}, []string{"method", "status_code", "path"})

	healthStatus = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "application_health_status",
		Help: "Health status of the application (1 for healthy, 0 for unhealthy)",
	})

	readinessStatus = prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "application_readiness_status",
		Help: "Readiness status of the application (1 for ready, 0 for not ready)",
	})
)

func init() {
	prometheus.MustRegister(httpRequestDuration, httpRequestCounter, healthStatus, readinessStatus)
}

func shouldExcludePath(path string) bool {
	excludedPaths := []string{"/", "/favicon.ico", "/metrics", "/health", "/ready"}
	for _, excludedPath := range excludedPaths {
		if path == excludedPath {
			return true
		}
	}
	return false
}

func TrackRequestDurationAndCount() gin.HandlerFunc {
	return func(c *gin.Context) {
		path := c.Request.URL.Path
		if shouldExcludePath(path) {
			c.Next()
			return
		}

		startTime := time.Now()
		c.Next()

		duration := time.Since(startTime).Seconds()
		statusCode := strconv.Itoa(c.Writer.Status())
		method := c.Request.Method

		httpRequestDuration.WithLabelValues(method, statusCode, path).Observe(duration)
		httpRequestCounter.WithLabelValues(method, statusCode, path).Inc()
	}
}

func SetHealthStatus(isHealthy bool) {
	if isHealthy {
		healthStatus.Set(1) // Set to 1 for healthy
	} else {
		healthStatus.Set(0) // Set to 0 for unhealthy
	}
}

func SetReadinessStatus(isReady bool) {
	if isReady {
		readinessStatus.Set(1) // Set to 1 for ready
	} else {
		readinessStatus.Set(0) // Set to 0 for not ready
	}
}

File: DeviceMS/WebApp/Middleware/prometheus/CustomMetrics.cs
using System.Diagnostics;
using Prometheus;

namespace WebApp.Middleware.prometheus;

public class CustomMetrics
{
    // Custom Prometheus metrics for ready and health checks
    public Gauge HealthCheckGauge { get; } = Metrics.CreateGauge(
        "application_health_status",
        "Health status of the application (1 for healthy, 0 for unhealthy)"
    );

    public Gauge ReadinessCheckGauge { get; } = Metrics.CreateGauge(
        "application_readiness_status",
        "Readiness status of the application (1 for healthy, 0 for unhealthy)"
    );
    
    // Custom Prometheus metrics for HTTP requests
    public Histogram HttpRequestDuration { get; } = Metrics.CreateHistogram(
        "http_request_duration_seconds",
        "Duration of HTTP requests in seconds",
        new HistogramConfiguration
        {
            LabelNames = new[] { "method", "status_code", "path" }
        }
    );

    public Counter HttpRequestCounter { get; } = Metrics.CreateCounter(
        "http_request_total",
        "Total count of HTTP requests",
        new CounterConfiguration
        {
            LabelNames = new[] { "method", "status_code", "path" }
        }
    );

    // Custom Metric for process resident memory in bytes
    public Gauge ProcessResidentMemoryBytes { get; } = Metrics.CreateGauge(
        "process_resident_memory_bytes",
        "Resident memory size of the process in bytes"
    );

    public CustomMetrics()
    {
        ProcessResidentMemoryBytes.Set(Process.GetCurrentProcess().WorkingSet64);
    }
}